controller/src/main/java/io/pravega/controller/store/stream/ZKGarbageCollector.java

```diff
@@ -0,0 +1,241 @@
+    private static final String BATCH_PATH = GC_ROOT + "/batch";
+    private static final String LEADER_PATH = GC_ROOT + "/leader";
+
+    private final ZKStoreHelper zkStoreHelper;
+    private final AtomicLong currentBatch;
+    private final AtomicReference<CompletableFuture<Void>> gcLoop;
+    private final CompletableFuture<Void> latch = new CompletableFuture<>();
+    private final Supplier<CompletableFuture<Void>> gcProcessing;
+    private final String gcName;
+    private final String leaderPath;
+    private final String batchPath;
+    private final AtomicReference<NodeCache> watch;
+    private final AtomicReference<LeaderLatch> leaderLatch;
+    // dedicated thread executor for GC which will not interfere with rest of the processing.
+    private final ScheduledExecutorService gcExecutor;
+    private final long periodInMillis;
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final CompletableFuture<Void> acquiredLeadership = new CompletableFuture<>();
+    
+    ZKGarbageCollector(String gcName, ZKStoreHelper zkStoreHelper, Supplier<CompletableFuture<Void>> gcProcessing, Duration gcPeriod) {
+        Preconditions.checkNotNull(zkStoreHelper);
+        Preconditions.checkNotNull(gcProcessing);
+        Preconditions.checkArgument(gcPeriod != null && !gcPeriod.isNegative());
+
+        this.currentBatch = new AtomicLong();
+        this.gcName = gcName;
+        this.leaderLatch = new AtomicReference<>();
+        this.leaderPath = String.format(LEADER_PATH, gcName);
+        this.watch = new AtomicReference<>();
+        this.batchPath = String.format(BATCH_PATH, gcName);
+        this.zkStoreHelper = zkStoreHelper;
+        this.gcProcessing = gcProcessing;
+        this.periodInMillis = gcPeriod.toMillis();
+        this.gcExecutor = Executors.newSingleThreadScheduledExecutor();
+        this.gcLoop = new AtomicReference<>();
+    }
+
+    @Override
+    protected void doStart() {
+        long quantized = computeNewBatch();
+        byte[] b = getBytes(quantized);
+        // 1. create znode for storing latest batch id. If the batch id exists, get the value from the store.
+        // We will later register watch on the path and keep receiving any changes to its value.
+        zkStoreHelper.createZNodeIfNotExist(batchPath, b)
+                .thenCompose(x -> zkStoreHelper.getData(batchPath))
+                .thenAccept(data -> {
+                    currentBatch.set(BitConverter.readLong(data.getData(), 0));
+                })
+                .thenRun(this::initialize)
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStarted();
+                    }
+                    latch.complete(null);
+                });
+    }
+
+    @Override
+    protected void doStop() {
+        latch.thenAccept(v -> {
+            if (gcLoop.get() != null) {
```

Comment:
there is a race between these get() calls. The value could change in between.

Reply:
Yes, there is a race in setting GC after start and if stop is called at the same time. Will try to fix it.